mol-loader/Pubchem.py

```diff
@@ -0,0 +1,69 @@
+"""
+Encapsulate reading and formatting of PubChem
+"""
+
+from Sdf import Sdf
+import json
+import os, os.path
+import rdkit.Chem
+
+class Pubchem(Sdf):
+    def __init__(self, path):
+        Sdf.__init__(self, path)
+
+    SYNONYM_PROPS = [
+        'PUBCHEM_IUPAC_OPENEYE_NAME',
+        'PUBCHEM_IUPAC_CAS_NAME',
+        'PUBCHEM_IUPAC_NAME',
+        'PUBCHEM_IUPAC_SYSTEMATIC_NAME',
+        'PUBCHEM_IUPAC_TRADITIONAL_NAME'
+    ]
+    SMILES_PROPS = [
+        'PUBCHEM_OPENEYE_CAN_SMILES',
+        'PUBCHEM_OPENEYE_ISO_SMILES'
+    ]
+
+    def mol_to_dict(self, mol):
+        """
+        Capture all the information from a PubChem molecule into a dictionary
+
+        mol_to_dict( (Pubchem)self, (rdkit.Chem.Mol)mol) -> dict
+        """
+        props = set(mol.GetPropNames())
+        # All the properties defined in the molecule, as-is
+        d = {p.lower() : self.cast(mol.GetProp(p)) for p in props}
+        d['id'] = 'https://pubchem.ncbi.nlm.nih.gov/compound/' + mol.GetProp('PUBCHEM_COMPOUND_CID')
+
+        smiles = [mol.GetProp(p) for p in self.SMILES_PROPS if p in props]
+        smiles.append(rdkit.Chem.MolToSmiles(mol)) # Original SMILES
+        # Attempt normalization; this may fail if there are ... oddities ...
+        # in the molecule.
+        rdkit_mol = rdkit.Chem.MolFromSmiles(rdkit.Chem.MolToSmiles(mol)) # Canonicalizes
+        if rdkit_mol is not None:
+            d['rdkit_smiles'] = rdkit.Chem.MolToSmiles(rdkit_mol)
+            smiles.append(d['rdkit_smiles'])
+        if smiles:
+            d['smiles'] = list(set(smiles))
+
+        synonyms = [mol.GetProp(p) for p in self.SYNONYM_PROPS if p in props]
+        if synonyms:
+            d['synonyms'] = list(set(synonyms))
+
+        if 'pubchem_compound_canonicalized' in d:
```

Comment:
alright - if it's confusing, I should change it to make it less confusing.
The goal is to fix up the types of some things, e.g. I'd love to get the CASE statements out of the ChEMBL SQL query and do the type fixing in Python.
The structure I have in mind is a dict from prop to value-translation-dict; the value-translation-dict, in this case, would be
```
{
    0 : False,
    1 : True
}
```